liveupdate: provide more user-friendly errors

milas opened the pull request:
Follow up to #5527.

The big addition here is to handle the case where `tar` is not
found in `$PATH`, which is very common when users are using a
`scratch`-built image or very minimal base image like the `distroless`
family.

How these manifest between the Docker updater (used for Compose +
K8s exposed in-cluster Docker socket) and exec updater (used for all
other K8s configs, generally meaning containerd) are slightly
different.

## Docker
### Permission Denied
<img width="1478" alt="dockerd_permissions" src="https://user-images.githubusercontent.com/841263/157087101-0bdc2eca-6d6c-4157-93e5-d1bb6dcf2445.png">

### No `tar` in PATH
<img width="1477" alt="dockerd_notfound" src="https://user-images.githubusercontent.com/841263/157087140-d1a10198-9bca-4c43-8e4b-0936b2a36d3b.png">

## Containerd
### Permission Denied
<img width="1459" alt="containerd_permissions" src="https://user-images.githubusercontent.com/841263/157087205-a83b4ea6-56bb-4917-8c42-7f9ef1333034.png">

### No `tar` in PATH
<img width="1475" alt="containerd_notfound" src="https://user-images.githubusercontent.com/841263/157087216-abcc1236-93c6-41a2-8d9d-081f729a680c.png">

nicksieger commented on internal/containerupdate/exec_updater.go:
```diff
@@ -50,9 +52,18 @@ func (cu *ExecUpdater) UpdateContainer(ctx context.Context, cInfo liveupdates.Co
 	if err != nil {
+		if exitCodeErr, ok := build.WrapCodeExitError(err, tarCmd); ok {
+			switch exitCodeErr.ExitCode {
+			case TarExitCodePermissionDenied:
+				return permissionDeniedErr(err)
+			case GenericExitCodeCannotExec:
+				return cannotExecErr(err)
```
Is it worth extracting this switch to a new function? Or moving to build.WrapCodeExitError()?

milas replied:
(see top-level comment - I tried to refactor for this + some other improvements)

milas commented:
@nicksieger I pushed a (shockingly chonky) commit (5a5afebd6) with some clean up

* Split concept of container exec errors from `RunStepFailure`
  * `RunStepFailure` is a generic wrapper error based on _where_ the failure occurred so that LU reconciler can react differently
  * This allows consistent/uniform handling of exec errors during things like the `tar` step to sync files and the actual `run()` steps execution
* Consolidated some error-handling logic in `containerupdate`
  * There's still a lot of duplication here because the K8s exec API + `client-go` doesn't give us as much as Docker, so we still have to peek at error messages in some cases
* Added a new test case for "not found in PATH" as well
* Swapped some logic to new stdlib error unwrapping and fix a couple issues here
* Remove redundant parts of errors (in some cases you could see the container ID or command multiple times)
* Fix mismatched capitalization in (wrapped) error messages

nicksieger reviewed (approved):
Sweet! Shows how error handling can be pretty verbose at times but it definitely makes the error handling logic clearer.
